minimal_high_density.md

```diff
@@ -0,0 +1,103 @@
+
+## Local smoothed density
+
+Local density used as an input for mortality is computed in `minimal.slim`
+by averaging over a Gaussian kernel with standard deviation `SX`.
+To approximate this, we simply smooth the "raw" gridded density by that same kernel:
+```
+DENSITY.smooth(SX * 3, "n", SX);
+```
+and then look up the local, smoothed density at the location of every individual:
+```
+competition = p1.spatialMapValue(DENSITY, inds.spatialPosition);
+```
+This could be made more precise by taking into account that in computing the "raw" density
+also involves some smoothing (the size of the grid squares).
+
+## Mate choice
+
+Mate choice in `minimal.slim` picks a nearby individual proportional to a weight
+assigned from a Gaussian kernel with standard deviation `SM`.
+To approximate this procedure, we choose a nearby location proportional to raw density
+multiplied by the same Gaussian kernel,
+and then return the nearest individual to that location:
+```
+mate_location = RAW_DENSITY.sampleNearbyPoint(individual.spatialPosition, 3 * SM, "n", SM);
+mate = i2.nearestNeighborsOfPoint(mate_location, p1, 1);
+```
+
+## Possible pitfalls
+
+The two lines in the "mate choice" snippet can each cause their own problems if care is not taken.
+
+First, the call to `sampleNearbyPoint` works by rejection sampling:
+it picks a location nearby from the provided kernel,
+and then retains the point with probability proportional to the value of the `RAW_DENSITY` map at that point.
+The way this is implemented means that if values of density in some parts of the map are much higher than in other parts of the map,
+this may do a large number of rejections.
+Concretely, if the density within a circle of radius `3 * SM` around a given individual
+is at most, say $10^{-6}$ times the maximum density over the entire map,
+then choosing a mate for that individual will require sampling millions of locations.
+The solution to this might be biological: first, identify individuals with sufficent possible mates
+to be able to mate successfully; and then only choose mates for those.
+
+Second, the call to `i2.nearestNeighborsOfPoint` will fail, clearly, if there *are* no neighbors.
+Since this call uses the interaction `i2`,
+which was set up at the start of the simulation 
+```
+initializeInteractionType(2, "xy", reciprocal=T, maxDistance=5/sqrt(K));
+```
+to have maximum interaction distance `5/sqrt(K)`.
+If the density is roughly $K$, this means there should be around 25 individuals in each circle of that radius;
+however, if density is sufficiently nonuniform, or if you are simulating some nonequilibrium situation,
+then this may fail.
+The code is robust to this:
+```
+if (mate.size())
+    subpop.addCrossed(individual, mate, count=rpois(1, FECUN));
+```
+means that if an individual has no neighbors, they will not reproduce,
+but this is over a much smaller distance than our nominal mating distance, `SM`,
+so we don't want this to happen very much.
```

Comment:
This needs a bit of rewriting.  "means that..." seems like it starts a sentence in the middle, and "this is over" and "we don't want this" both use "this" in an ambiguous way that I am honestly unable to puzzle out.